contracts/forwarder/Eip712Forwarder.sol

```diff
@@ -0,0 +1,123 @@
+// SPDX-License-Identifier:MIT
+pragma solidity ^0.6.2;
+pragma experimental ABIEncoderV2;
+
+import "@openzeppelin/contracts/cryptography/ECDSA.sol";
+import "./IForwarder.sol";
+// TODO now: move getError to the forwarder folder
+import "../utils/GsnUtils.sol";
+
+contract Eip712Forwarder is IForwarder {
+    using ECDSA for bytes32;
+
+    string public constant GENERIC_PARAMS = "address to,bytes data,address from,uint256 nonce,uint256 gas";
+
+    mapping(bytes32 => bool) public typeHashes;
+
+    // Nonces of senders, used to prevent replay attacks
+    mapping(address => uint256) private nonces;
+
+    function getNonce(address from)
+    public view override
+    returns (uint256) {
+        return nonces[from];
+    }
+
+    function verify(
+        ForwardRequest memory req,
+        bytes32 domainSeparator,
```

Comment:
why is the domain separator not hardcoded in the forwarder?
Allowing Forwarder extension to change it does not seem like a good idea as this might confuse user

Reply:
Depends what we define as "domain". 
I was thinking of "Domain" as the relaying service in use (in our case, its "GSN"). It has nothing to do with the client's signature on the request.

Reply:
ok, I see